espnet2/train/trainer.py

```diff
@@ -127,15 +162,11 @@ def train_one_epoch(
                 loss, stats, weight = data_parallel(
                     model, (), range(ngpu), module_kwargs=batch
                 )
-                # Weighted averaging of loss from torch-data-parallel
-                loss = (loss * weight.to(loss.dtype)).sum(0) / weight.sum()
-                stats = {
-                    k: (v * weight.to(v.dtype)).sum(0) / weight.sum()
-                    if v is not None
-                    else None
-                    for k, v in stats.items()
-                }
-                weight = weight.sum()
+            if ngpu > 1 or options.distributed:
+                # Weighted averaging
+                loss = (loss * weight).sum() / weight.sum()
+                stats, weight = recursive_average(stats, weight, options.distributed)
+
```

Comment:
About distributed training: In pytorch distributed training, the training process can be done transparently with `DistributedDataParallel`, just like `DataParallel`, because `all-reduce` operations are hidden in the wrapper class when `backward()`.

The `stats` need manual all reduce operation since it's not relevant to `backward()` operation, so I defined `recursive_average()` here.

Reply:
I think it is better to add NOTE about your comments.